File: ios/Simbi/General/_SMBChatButton.m
//
//  SMBChatButton.m
//  Simbi
//
//  Created by flynn on 5/29/14.
//  Copyright (c) 2014 SimbiSocial. All rights reserved.
//

#import "_SMBChatButton.h"

#import "SMBAppDelegate.h"


@interface _SMBChatButton ()

@property (nonatomic, strong) UIButton *responseButton;
@property (nonatomic, strong) UIView *buttonView;
@property (nonatomic, strong) UIView *notificationView;
@property (nonatomic, strong) UILabel *notificationLabel;

@property (nonatomic) CGRect notificationHiddenFrame;
@property (nonatomic) CGRect notificationNormalFrame;
@property (nonatomic) CGRect notificationWideFrame;

@end


@implementation _SMBChatButton

- (id)initWithTarget:(id)target action:(SEL)action
{
    self = [super init];
    
    if (self)
    {
        _notificationHiddenFrame = CGRectMake(2+13/2.f, 13+13/2.f, 0, 0);
        _notificationNormalFrame = CGRectMake(2, 13, 13, 13);
        _notificationWideFrame   = CGRectMake(0, 13, 17, 13);
        
        
        _buttonView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 44, 44)];
        
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(8, 8, 44-16, 44-16)];
        [imageView setImage:[UIImage imageNamed:@"Chat_Icon.png"]];
        [_buttonView addSubview:imageView];
        
        if ([SMBUser currentUser].hasNewMessage || [[SMBAppDelegate instance] isAtHomeOrChat])
        {
            if ([SMBUser currentUser].unreadMessageCount.intValue < 10)
                _notificationView = [[UIView alloc] initWithFrame:_notificationNormalFrame];
            else
                _notificationView = [[UIView alloc] initWithFrame:_notificationWideFrame];
        }
        else
            _notificationView = [[UIView alloc] initWithFrame:_notificationHiddenFrame];
        [_notificationView setBackgroundColor:[UIColor simbiRedColor]];
        [_notificationView.layer setCornerRadius:_notificationView.frame.size.height/2.f];
        [_notificationView.layer setBorderColor:[UIColor whiteColor].CGColor];
        [_notificationView.layer setBorderWidth:0.5f];
        [_notificationView.layer setMasksToBounds:YES];
        [_buttonView addSubview:_notificationView];
        
        _notificationLabel = [[UILabel alloc] initWithFrame:_notificationWideFrame];
        if ([SMBUser currentUser].unreadMessageCount.intValue > 0)
            [_notificationLabel setText:[SMBUser currentUser].unreadMessageCount.stringValue];
        else
            [_notificationLabel setAlpha:0.f];
        [_notificationLabel setTextColor:[UIColor whiteColor]];
        [_notificationLabel setFont:[UIFont simbiBoldFontWithSize:9.f]];
        [_notificationLabel setTextAlignment:NSTextAlignmentCenter];
        [_buttonView addSubview:_notificationLabel];
        
        _responseButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_responseButton setFrame:CGRectMake(2, 2, 40, 40)];
        [_responseButton addTarget:target action:action forControlEvents:UIControlEventTouchUpInside];
        [_buttonView addSubview:_responseButton];
        
        [self setCustomView:_buttonView];
        
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(showChatNotification:) name:kSMBNotificationShowChatIcon object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(hideChatNotification:) name:kSMBNotificationHideChatIcon object:nil];
    }
    
    return self;
}


- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


- (void)showChatNotification:(NSNotification *)notification
{
    if (![[SMBAppDelegate instance] isAtHomeOrChat])
    {
        [_notificationLabel setText:[SMBUser currentUser].unreadMessageCount.stringValue];
        
        if (_notificationLabel.text.length > 2)
            [_notificationLabel setText:@"99+"];
        
        [self animateNotificationViewIn];
    }
}


- (void)hideChatNotification:(NSNotification *)notification
{
    [_notificationLabel setText:[SMBUser currentUser].unreadMessageCount.stringValue];
    
    if (_notificationLabel.text.length > 2)
        [_notificationLabel setText:@"99+"];
    
    [self animateNotificationViewOut];
}


- (void)animateNotificationViewIn
{
    [UIView animateWithDuration:0.33f
                     animations:^{
                         if (_notificationLabel.text.length > 1)
                             [_notificationView setFrame:_notificationWideFrame];
                         else
                             [_notificationView setFrame:_notificationNormalFrame];
                         [_notificationView.layer setCornerRadius:_notificationView.frame.size.height/2.f];
                     }
                     completion:^(BOOL finished) {
                         [UIView animateWithDuration:0.33f
                                          animations:^{
                                              [_notificationLabel setAlpha:1.f];
                                          }];
                     }
    ];
}


- (void)animateNotificationViewOut
{
    [UIView animateWithDuration:0.33f
                     animations:^{
                         [_notificationLabel setAlpha:0.f];
                         
                     }
                     completion:^(BOOL finished) {
                         [UIView animateWithDuration:0.33f
                                          animations:^{
                                              [_notificationView setFrame:_notificationHiddenFrame];
                                          }];
                     }
    ];
}


@end

File: ios/Simbi/General/SMBChatButton.swift
//
//  SMBChatButton.swift
//  Simbi
//
//  Created by flynn on 10/23/14.
//  Copyright (c) 2014 SimbiSocial. All rights reserved.
//

import UIKit


class SMBChatButton: UIButton {
    
    var notificationHiddenFrame = CGRectZero
    var notificationNormalFrame = CGRectZero
    var notificationWideFrame   = CGRectZero
    
    let borderView = UIView()
    let notificationView = UIView()
    let notificationLabel = UILabel()
    
    override var frame: CGRect {
        
        didSet {
            
            self.layer.cornerRadius = frame.height/2
            
            borderView.frame = CGRectMake(0, 0, frame.width, frame.height)
            borderView.layer.cornerRadius = borderView.frame.height/2
            
            // Frames for the different notification states
            
            notificationHiddenFrame = CGRectMake(2+13/2, frame.height-13/2,  0,  0)
            notificationNormalFrame = CGRectMake(     0, frame.height-13  , 13, 13)
            notificationWideFrame   = CGRectMake(     0, frame.height-13  , 13, 13)
            //notificationWideFrame   = CGRectMake(    -2, frame.height-13  , 17, 13)
            
            // Adjust frame for subviews
            
            if (SMBUser.currentUser() != nil && SMBUser.currentUser().hasNewMessage) || SMBAppDelegate.instance().isAtHomeOrChat() {
                
                if SMBUser.currentUser().unreadMessageCount != nil {
                    
                    if SMBUser.currentUser().unreadMessageCount != nil {
                        
                        if SMBUser.currentUser().unreadMessageCount.intValue == 0 {
                            notificationView.frame = notificationHiddenFrame
                            notificationView.alpha = 0
                        }
                        else if SMBUser.currentUser().unreadMessageCount.intValue < 10 {
                            notificationView.frame = notificationNormalFrame
                        }
                        else {
                            notificationView.frame = notificationWideFrame
                        }
                    }
                    else { notificationView.frame = notificationWideFrame }
                }
                else {
                    notificationView.frame = notificationWideFrame
                }
            }
            else {
                notificationView.frame = notificationHiddenFrame
            }
            notificationView.layer.cornerRadius = notificationView.frame.height/2
            
            notificationLabel.frame = notificationWideFrame
        }
    }
    
    
    // MARK: - View Lifecycle
    
    required init(coder aDecoder: NSCoder) { fatalError("init(coder:) has not been implemented") }
//    
//    convenience init() {
//        self.init()
//        loadView()
//    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        loadView()
    }
    
    
    func loadView() {
        
        self.setImage(UIImage(named: "chat_icon"), forState: .Normal)
        self.tintColor = UIColor.simbiBlueColor()
        
        
        // Set up subviews
        
        borderView.layer.borderColor = UIColor.whiteColor().CGColor
        borderView.layer.borderWidth = 1
        borderView.hidden = true
        borderView.userInteractionEnabled = false
        self.addSubview(borderView)
        
        notificationView.backgroundColor = UIColor.simbiRedColor()
        notificationView.layer.cornerRadius = notificationView.frame.height/2
        notificationView.layer.masksToBounds = true
        notificationView.userInteractionEnabled = false
        self.addSubview(notificationView)
        
        if SMBUser.currentUser() != nil &&
           SMBUser.currentUser().unreadMessageCount != nil &&
           SMBUser.currentUser().unreadMessageCount.intValue > 0 {
            
            notificationLabel.text = SMBUser.currentUser().unreadMessageCount.stringValue
        }
        else {
            notificationLabel.alpha = 0
        }
        notificationLabel.textColor = UIColor.whiteColor()
        notificationLabel.font = UIFont.simbiBoldFontWithSize(9)
        notificationLabel.textAlignment = .Center
        notificationLabel.userInteractionEnabled = false
        //self.addSubview(notificationLabel)
        
        
        // Subscribe to notifications
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "showChatNotification:", name: kSMBNotificationShowChatIcon, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "hideChatNotification:", name: kSMBNotificationHideChatIcon, object: nil)
    }

    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    
    // MARK: - Notification Handling
    
    func showChatNotification(notification: NSNotification) {
        
        if SMBAppDelegate.instance().isAtHomeOrChat() {
            
            notificationLabel.text = SMBUser.currentUser().unreadMessageCount.stringValue
            
            
            if (notificationLabel.text?.characters.count) > 2 {
                notificationLabel.text = "99+"
            }
            
            animateNotificationViewIn()
        }
    }
    
    
    func hideChatNotification(notification: NSNotification) {
        
        if SMBUser.currentUser().unreadMessageCount != nil {
            notificationLabel.text = SMBUser.currentUser().unreadMessageCount.stringValue
        }
        else {
            notificationLabel.text = ""
        }
        
        if (notificationLabel.text?.characters.count) > 2 {
            notificationLabel.text = "99+"
        }
        
        animateNotificationViewOut()
    }
    
    
    // MARK: Public Methods
    
    func showBackground(shouldShow: Bool) {
        
        borderView.hidden = !shouldShow
        
        if shouldShow {
            self.backgroundColor = UIColor.simbiWhiteColor().colorWithAlphaComponent(0.9)
            self.layer.shadowOffset = CGSizeMake(1, 1)
            self.layer.shadowColor = UIColor.blackColor().CGColor
            self.layer.shadowOpacity = 0.33
        }
        else {
            self.backgroundColor = UIColor.clearColor()
            self.layer.shadowOpacity = 0
        }
    }
    
    
    // MARK: Private Methods
    
    private func animateNotificationViewIn() {
        
        UIView.animateWithDuration(0.33, animations: { () -> Void in
            
            self.notificationView.frame = (self.notificationLabel.text?.characters.count) > 1 ?
                self.notificationWideFrame : self.notificationNormalFrame

            self.notificationView.layer.cornerRadius = self.notificationView.frame.height/2
            
        }) { (Bool) -> Void in
            
            UIView.animateWithDuration(0.33, animations: { () -> Void in
                self.notificationLabel.alpha = 1
            })
        }
    }
    
    
    private func animateNotificationViewOut() {
        
        UIView.animateWithDuration(0.33, animations: { () -> Void in
            
            self.notificationLabel.alpha = 0
            
        }) { (Bool) -> Void in
            
            UIView.animateWithDuration(0.33, animations: { () -> Void in
                self.notificationView.frame = self.notificationHiddenFrame
            })
        }
    }
}
